### PR TITLE
Fix DoImmutableMatrix for finite fields

### PR DESCRIPTION
**Pull request by frankluebeck**
DoImmutableMatrix wrongly assumed that finite fields of order q <= 256
are represented as GF(q) and applied the NC versions of ConvertToMatrixRep
to matrices over such fields which produced corrupted objects. E.g.:
```
gap> f1 := AlgebraicExtension(GF(3), CyclotomicPolynomial(GF(3), 5));;
gap> a := RootOfDefiningPolynomial(f1);;
gap> mat := [[a]];;
gap> DoImmutableMatrix(f1, mat, false);
< immutable compressed matrix 1x393286524 over GF(26) >
```

### CODE REVIEW

**Comment by frankluebeck** [comment]
This pull request resolves the bug reported in issue "Bug in iterated algebraic extensions over finite fields"  #1457.

File: lib/vecmat.gi [hunk 1]
```diff
@@ -1411,12 +1411,12 @@ local sf, rep, ind, ind2, row, i,big,l;
   else
     if not IsField(field) then
       # not a field
-      return Immutable(matrix);
+      return matrix;
```
Comment (fingolfin) [inline lib/vecmat.gi]: Why this change? It is not mentioned in your commit message, and at a quick glance, seems wrong.

File: lib/vecmat.gi [hunk 2]
```diff
@@ -1473,18 +1473,20 @@ local sf, rep, ind, ind2, row, i,big,l;
   fi;
 
   # this can only happen if not big
-  for i in ind do
-    ConvertToVectorRepNC(matrix[i],sf);
-  od;
+  if not big then
```
Comment (fingolfin) [inline lib/vecmat.gi]: Why this change? If `big` is true, then nothing gets addd to `ind` anyway.

File: lib/vecmat.gi [hunk 3]
```diff
@@ -1411,12 +1411,12 @@ local sf, rep, ind, ind2, row, i,big,l;
...
     if not IsField(field) then
       # not a field
-      return Immutable(matrix);
+      return matrix;
     fi;
     sf:=Size(field);
   fi;
 
-  big:=sf>256 or sf=0;
+  big:=sf>256 or sf=0 or (sf <= 256 and not field=GF(sf));
```
Comment (fingolfin) [inline lib/vecmat.gi]: The `sf <= 256` is redundant, due to the checks before it.

I am also a bit wary of writing `field=GF(sf)`, and its performance impact. Of course if this is called a lot with the same value, it won't matter due to `GFCACHE`; but if `sf` alternates between different field sizes. But then again, I am perhaps worrying too much (sure, there is a primality check, and for small `sf` it's fast, and then we end up using `GALOIS_FIELDS`, the second cache for small galois fields...)

Anyway, how about using `IsFFECollection`, then this is simply avoided. I.e.:

`big:=sf>256 or sf=0 or not IsFFECollection(field);`

File: lib/vecmat.gi (hunk 1 above)
Comment (frankluebeck) [inline lib/vecmat.gi]: I have not touched this line. Seems that this very recent change got lost when I ported my fix to the latest  master branch. I will add back that `Immutable`.

I don't understand what this case means. What are valid arguments for this function?

File: lib/vecmat.gi (hunk 3 above)
Comment (frankluebeck) [inline lib/vecmat.gi]: I didn't find a filter that would test what I want (and thought that this is not performance critical because the GF(q)'s would only be generated for small q).
But reading the documentation of `IsFFECollection` I guess that this is a correct test to decide if the `Convert...` and `CONV...` functions are applicable in the following code.

Will change as suggested, thanks for the hint.

But the function also seems to support an integer as `field` argument. Maybe both variants are not correct?

File: lib/vecmat.gi (hunk 2 above)
Comment (frankluebeck) [inline lib/vecmat.gi]: This seems to be the easiest way to make sure that the `Convert...NC` function is not called if `big=true`. Actually, I did not try to understand or check the code that determined `ind`, because I don't know what exactly this function is supposed to do.

This pull request is only meant as quick and local fix (to make  `AlgebraicExtension` work for me).

I'm happy if someone who knows better this code provides a different fix (and maybe adds some explanations about the meaning of this code).

Comment (fingolfin) [inline lib/vecmat.gi]: OK, fair enough. I also find the code hard to understand (had to resd it recently foe the ImmutableVector work). I'll takr a other look at this later, and either merge this PR, or do as you suggest and make a new one which rewrites the function with better comments etc.

**Comment by markuspf** [comment]
This has been fixed by #1504.
